examples/esp32/xively_demo/main/main.c

```diff
@@ -0,0 +1,293 @@
+    user_data_printf( &user_config );
+    return 0;
+#endif
+
+    /* Read user data from flash */
+    if ( 0 > user_data_copy_from_flash( &user_config ) )
+    {
+        printf( "\n[ERROR] trying to copy user data from flash. Abort" );
+        return -1;
+    }
+    printf( "\nUser data retrieved from flash:" );
+    user_data_printf( &user_config );
+
+    /* Wait for a button press for 80*50 ms while flashing the LED. If the button is
+     * pressed, clear the contents of user_config to force the provisioning process */
+    for ( int i = 80; i > 0; i-- )
+    {
+        if ( -1 != io_await_gpio_interrupt( 50 ) )
+        {
+            printf( "\nButton pressed - Initializing provisioning process" );
+            provisioning_bootmode_selected = 1;
+            break;
+        }
+        io_led_set( i % 2 ); /* Toggle LED */
+    }
+    io_led_off();
+
+    /* If the data retrieved from NVS is missing any fields, start provisioning */
+    if ( provisioning_bootmode_selected || ( 0 > user_data_is_valid( &user_config ) ) )
+    {
+        io_led_on();
+        if ( 0 > provisioning_gather_user_data( &user_config ) )
+        {
+            printf( "\nDevice provisioning [ERROR]. Abort" );
+            io_led_off();
+            return -1;
+        }
+    }
+    io_led_off();
+    return 0;
+}
+
+void app_gpio_interrupts_handler_task( void* param )
+{
+    const uint32_t IO_BUTTON_WAIT_TIME_MS = 500;
+    int button_input_level = -1;
+    int virtual_switch = 0; /* Switches between 1 and 0 on each button press */
+    while ( 1 )
+    {
+        if ( -1 !=
+             ( button_input_level = io_await_gpio_interrupt( IO_BUTTON_WAIT_TIME_MS ) ) )
+        {
+            virtual_switch = !virtual_switch;
+            printf( "\nButton pressed! Virtual switch [%d]", virtual_switch );
+            xif_publish_button_state( virtual_switch );
+        }
+        taskYIELD();
+    }
+}
+
+void xif_recv_mqtt_msg_callback( const xi_sub_call_params_t* const params )
+{
+    printf( "\nNew MQTT message received!" );
+    printf( "\n\tTopic: %s", params->message.topic );
+    printf( "\n\tMessage size: %d", params->message.temporary_payload_data_length );
+    printf( "\n\tMessage payload: " );
+    for ( size_t i = 0; i < params->message.temporary_payload_data_length; i++ )
+    {
+        // printf( "0x%02x ", params->message.temporary_payload_data[i] );
+        printf( "%c", ( char )params->message.temporary_payload_data[i] );
+    }
+    if( 0 == strcmp( params->message.topic, xif_mqtt_topics.led_topic ) )
+    {
+        if ( 0 == params->message.temporary_payload_data_length )
+            io_led_off();
+        else if ( '0' == params->message.temporary_payload_data[0] )
+            io_led_off();
+        else
+            io_led_on();
+    }
+}
+
+void app_main( void )
+{
+    /* Initialize Non-Volatile Storage */
+    if ( 0 > io_init() )
+    {
+        while ( 1 )
+            vTaskDelay( 1000 / portTICK_PERIOD_MS );
```

Comment:
Do these wait forever states deserve printfs first?  I'm not sure if printfs even work yet, but i'm asking anyway. :)